lxd/instance/drivers/driver_qemu.go

```diff
@@ -6579,6 +6579,22 @@ func (d *qemu) MigrateSend(args instance.MigrateSendArgs) error {
 		return fmt.Errorf("Failed receiving migration offer response: %w", err)
 	}
 
+	// If we have unknown fields in the protobuf message, then we received
+	// unexpected data. This will only happen if we were sent a migration.MigrationControl
+	// instead of the expected migration.MigrationHeader. This implies that we have failed on
+	// the target before we were able to send the response header.
+	unknownFields := offerHeader.ProtoReflect().GetUnknown()
+	if len(unknownFields) > 0 {
+		msg, err := shared.GetFirstUnknownField(unknownFields)
```

Comment:
Relying on the first field feels a bit arbitrary. I wonder if we can make the underlying `ControlReceive` a bit smarter to avoid this?

`ControlReceive` calls [`ProtoRecv`](https://github.com/canonical/lxd/blob/main/lxd/migration/wsproto.go#L14-L39) which unmarshals the data into the provided type (which it interprets as a `proto.Message`)

Could we split up `ProtoRecv` into two parts:
```go
// Reads from the conn and returns the received []byte.
func ProtoRead(ws *websocket.Conn) ([]byte, error)
```
and
```go
// Unmarshals a []byte into a proto.Message.
 func ProtoParse(buf []byte, msg proto.Message) error
```
Then I _think_ we could do something like this:
```go
func recv(m proto.Message) error {
    conn, err := c.conns[api.SecretNameControl].WebSocket(context.TODO())
    if err != nil {
        return fmt.Errorf("Control connection not initialized: %w", err)
    }

    // Get the underlying type of the proto.Message.
    t := reflect.TypeOf(m)
    if t.Kind() == reflect.Ptr {
        t = t.Elem()
    }

    // Read from the connection.
    buf, err := migration.ProtoRead(conn)
    if err != nil {
        return err
    }

    // If we are expecting a MigrationHeader, first check if we received a MigrationControl
    // which contains error messages indicating that we need to stop.
    if t == reflect.TypeOf(migration.MigrationHeader{}) {
        ctl := &migration.MigrationControl{}
        err := migration.ProtoParse(buf, ctl)
        if err != nil {
            return err
        }

        // Check if the fields are non-nil, indicating that we actually did receive a MigrationControl.
        // If Success if non-nil and false, then we received an error.
        if ctl.Success != nil && ctl.Message != nil && !ctl.GetSuccess() {
            return fmt.Errorf(ctl.GetMessage())
        }
    }

    // Properly parse into the correct type before returning.
    return migration.ProtoParse(buf, m)
}
```

Reply:
In particular, I have no idea if this block is correct:
```go
 // Check if the fields are non-nil, indicating that we actually did receive a MigrationControl.
        // If Success if non-nil and false, then we received an error.
        if ctl.Success != nil && ctl.Message != nil && !ctl.GetSuccess() {
            return fmt.Errorf(ctl.GetMessage())
        }
```

Reply:
Oh, you can actually simplify out the direct `reflect` package usage by just making a type assertion:
```go
_, ok := m.(*migration.MigrationHeader)
if ok {
    ctl := &migration.MigrationControl{}
    err := migration.ProtoParse(buf, ctl)
    if err != nil {
        return err
    }
 
    // Check if the fields are non-nil, indicating that we actually did receive a MigrationControl.
    // If Success if non-nil and false, then we received an error.
    if ctl.Success != nil && ctl.Message != nil && !ctl.GetSuccess() {
        return fmt.Errorf(ctl.GetMessage())
    }
}
```

Reply:
I really appreciate this! I've pushed new changes to use a fix along these lines instead of relying on unknown fields.